core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,1099 @@
+      'html content is null', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', null, null, []), 1, 0, 'default_outcome'
+      );
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+      let alertsServiceSpy =
+        spyOn(alertsService, 'addWarning').and.returnValue();
+      spyOn(expressionInterpolationService, 'processHtml')
+        .and.callFake((html, envs) => {
+          return html;
+        });
+
+      singleQuestionBackendDict.question_state_data
+        .interaction.default_outcome.feedback.html = null;
+      questionPlayerEngineService.init(
+        [singleQuestionBackendDict], successHandler, failHandler);
+
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(alertsServiceSpy)
+        .toHaveBeenCalledWith('Expression parsing error.');
+    });
+
+    it('should show warning message given that question ' +
+      'html content is null', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+
+      singleQuestionBackendDict.question_state_data
+        .content.html = null;
+      let sampleQuestion = questionObjectFactory.createFromBackendDict(
+        singleQuestionBackendDict);
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+      let alertsServiceSpy =
+        spyOn(alertsService, 'addWarning').and.returnValue();
+      spyOn(questionPlayerEngineService, 'init').and.callFake(() => {
+        questionPlayerEngineService.addQuestion(sampleQuestion);
+      });
+
+      spyOn(expressionInterpolationService, 'processHtml')
+        .and.callFake((html, envs) => {
+          return html;
+        });
+
+      questionPlayerEngineService.init(
+        [singleQuestionBackendDict], successHandler, failHandler);
+      questionPlayerEngineService.setCurrentIndex(0);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(alertsServiceSpy)
+        .toHaveBeenCalledWith('Expression parsing error.');
+    });
+
+    it('should return true if multiple questions exists', () => {
```

Comment:
exists --> exist

Also I don't know what "it" is here. Please fix the description.  Remember that "it" refers to the thing being "described".

Similar comments for the descriptions of most of the tests below. I'm going to stop reviewing here, please make sure all test descriptions are really clear, you're testing interface not implementation, there are pre-checks where needed, etc. before resending for review.

Reply:
done.